<a id="start"></a>
## Übung 1: Bibliothek
Eine Bibliothek speichert ihre vorhandenen Exemplare mithilfe einer lokalen Kennung.
Zu jeder lokalen Kennung gehört eine ISBN, die für jedes Exemplar eindeutig identifiziert,
um welches Buch es sich handelt. Zusätzlich speichert die Bibliothek ihre Nutzer und welche Bücher
sie jeweils ausgeliehen haben. In den folgenden Aufgaben wollen wir typische Anfragen in
Relationaler Algebra an diese Bibliotheksdatenbank stellen.


#### ER-Diagramm

![](../resources/04_relationale_algebra/bibliothek-schema.png)

In [1]:
from reframe import Relation

Benutzer = Relation('../resources/04_relationale_algebra/Benutzer.csv')
Leiht_aus = Relation('../resources/04_relationale_algebra/Leiht_aus.csv')
Exemplar = Relation('../resources/04_relationale_algebra/Exemplar.csv')
Buch = Relation('../resources/04_relationale_algebra/Buch.csv')

print(Benutzer)
print(Leiht_aus)
print(Exemplar)
print(Buch)

   Bid Vorname
0    1   Babsi
1    2     Tad
2    3     Odo
3    4   Laila
4    5     Pam
   Bid Signatur
0    1    KS345
1    5    ST220
2    1   JLb 12
3    2    ST270
4    5    ST270
  Signatur     ISBN
0    S2710  0-13-1X
1    ST270  0-13-1X
2   JLb 12  3-540-X
3    ST220  0-544-X
4    KS345   37020X
      ISBN                           Titel Sprache
0  0-13-1X                Database Systems      en
1  3-540-X  Wein : verstehen und geniessen      de
2  0-544-X  Big data : a revolution that …      en
3   37020X       Bierbrauen fuer jedermann      de
4   38266X     Data Warehouse Technologien      de


1. Welche englischsprachigen Bücher (Titel) gibt es?

2. Welche Benutzer haben noch nie ein Buch ausgeliehen?

3. Welche Benutzer (Vorname) haben ein Exemplar mit Buchtitel „Bierbrauen fuer jedermann“ ausgeliehen?

4. Alle Benutzer (Vorname), die mindestens zwei Exemplareausgeliehen haben.

## Übung 2: Minimale und maximale Ausgabekardinalität von Operatoren

* Gegeben seien zwei kompatible Relationen R(𝐵,𝐶) und S(𝐵,𝐶)
* Die Relationen enthalten jeweils |R| und |S| Tupel
* Welche Aussage können wir über |𝑆 ∪ 𝑇| treffen (Mengen)?
    - Maximal?
    - Minimal?
* Welche Aussage können wir über |𝑆 ∪ 𝑇| treffen (Multimengen)?
    - Maximum?
    - Minimum?

## Übung 3: Regatta-Informationssystem

Formulieren Sie die folgenden textuellen Anfragen mit Hilfe der Relationen-Algebra.

#### ER-Diagramm

![](../resources/04_relationale_algebra/regatta-schema.png)

In [2]:
Bootsklasse = Relation('../resources/04_relationale_algebra/Bootsklasse.csv')
Wettfahrt = Relation('../resources/04_relationale_algebra/Wettfahrt.csv')
Platzierung = Relation('../resources/04_relationale_algebra/Platzierung.csv')
Teilnehmer = Relation('../resources/04_relationale_algebra/Teilnehmer.csv')

print(Bootsklasse)
print(Wettfahrt)
print(Platzierung)
print(Teilnehmer)

      Klasse Bauart
0      Pirat   Holz
1  Folkeboot   Holz
2     H-Boot    GFK
3   Optimist    GFK
   FahrtNr                      Name    Datum    Zeit
0        1                 Moorpokal  18.6.03  10:00h
1        2             Herbstmeister  16.9.03  14:00h
2        3  Franz Huber Gedenk Preis  15.5.03  14:00h
3        4               Blaues Band  29.5.03  10:00h
     SegelNr  Wettfahrt Platz
0   GER 4318          1     1
1   GER 4318          2     2
2   GER 4995          1     2
3   GER 4995          2     1
4   GER 5703          4     2
5   GER 5107          4     1
6   GER 5505          4     3
7   GER 1393          3     1
8   GER 3876          3     4
9   GER 4309          3     3
10  GER 4833          3   dnf
11  GER 5503          3     2
     SegelNr                Name Bootsklasse Baujahr        Farbe     Eigner
0   GER 4318               Marie       Pirat    1992         Blau  D. Dummer
1   GER 4995             Celeste       Pirat    1991          Rot  S. Schott
2   GER 5

1. Aus welchem Material wurde das Boot mit der Segelnummer „GER 4318“ gebaut und welcher Klasse gehört es an?

2. Wann starteten die Teilnehmer am „29.5.03“ und welche Boote der Klasse „Optimist“ nahmen an dieser Regatta teil?

3. An welchen Wettfahrten haben ausschließlich Holzboote teilgenommen?

4. Welche Bootsklassen nehmen am „Franz Huber Gedenk Preis“ oder am „Moorpokal“ teil?

5. Welche Boote konnten das Rennen um das „Blaue Band“ nicht beenden („dnf“)?

6. Wie viele Boote haben am „Franz Huber Gedenk Preis“ teilgenommen?

7. An welchem Rennen nahm das älteste Boot im Starterfeld teil?


## Übung 4: Semantisches Verständnis
Geben Sie das Ergebnis und die umgangssprachliche Formulierung der folgenden Relationen Algebra Ausdrücke an:

![title](../resources/04_relationale_algebra/Ausdruck1.png)


![title](../resources/04_relationale_algebra/Ausdruck2.png)

![title](../resources/04_relationale_algebra/Ausdruck3.png)

